src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,212 @@
+        //   void LF1<T>()
+        //   {
+        //       void LF2<G>()
+        //       {
+        //           void LF3()
+        //           {
+        //               Func<T> d = SomeMethod<T>;
+        //               static void LF4 () { Func<T> d = SomeMethod<T>; }
+        //           }
+        //           
+        //           void LF5()
+        //           {
+        //               Func<T> d = SomeMethod<T>;
+        //           }
+        //       }
+        //   }
+        //
+        // In the above case, only one cached delegate is necessary, and it could be assigned to the container 'owned' by LF1.
+
+        if (!TryGetOwnerFunction(_factory.CurrentFunction, delegateType, targetMethod, out var ownerFunction))
+        {
+            var typeCompilationState = _factory.CompilationState;
+            container = typeCompilationState.ConcreteDelegateCacheContainer;
+
+            if (container is { })
+            {
+                return container;
+            }
+
+            container = new DelegateCacheContainer(typeCompilationState.Type, generation);
+            typeCompilationState.ConcreteDelegateCacheContainer = container;
+        }
+        else
+        {
+            var containers = _genericCacheContainers ??= new Dictionary<MethodSymbol, DelegateCacheContainer>();
+
+            if (containers.TryGetValue(ownerFunction, out container))
+            {
+                return container;
+            }
+
+            container = new DelegateCacheContainer(ownerFunction, _topLevelMethodOrdinal, containers.Count, generation);
+            containers.Add(ownerFunction, container);
+        }
+
+        _factory.AddNestedType(container);
+
+        return container;
+    }
+
+    private static bool TryGetOwnerFunction(MethodSymbol currentFunction, TypeSymbol delegateType, MethodSymbol targetMethod, [NotNullWhen(true)] out MethodSymbol? ownerFunction)
+    {
+        if (targetMethod.MethodKind == MethodKind.LocalFunction)
+        {
+            // Local functions can use type parameters from their enclosing methods!
+            //
+            // For example:
+            //   void Test<T>()
+            //   {
+            //       var t = Target<int>;
+            //       static object Target<V>() => default(T);
+            //   }
+            //
+            // Therefore, without too much analysis, we select the closest generic enclosing function as the cache container owner.
+
+            for (Symbol? enclosingSymbol = currentFunction; enclosingSymbol is MethodSymbol enclosingMethod; enclosingSymbol = enclosingSymbol.ContainingSymbol)
+            {
+                if (enclosingMethod.Arity > 0)
+                {
+                    ownerFunction = enclosingMethod;
+                    return true;
+                }
+            }
+
+            ownerFunction = null;
+            return false;
+        }
+
+        var usedTypeParameters = PooledHashSet<TypeParameterSymbol>.GetInstance();
```

Comment:
nit: consider asserting on MethodKind (we can only get Ordinary here, right?)

Reply:
Good question. I assumed `Ordinary` here. Is it possible ending up `EventAdd` or `PropertyGet`? Let me play a little bit and get back here.

Reply:
@jcouv

> consider asserting on MethodKind (we can only get Ordinary here, right?)

 I am not sure if this assert adds any value. I think that for the purpose of this component, the kind of the target method is not significant. We are not relying on the fact anywhere.

Reply:
It adds to readability in my opinion (we know what we're dealing with).

Reply:
> It adds to readability in my opinion (we know what we're dealing with).

I don't think I can agree with this. There is a whole bunch of unrelated things that we can assert here. But we are not doing this. In my opinion this assert hurts readability. It confuses the reader, making him think that some how doing all this here for other methods is going to be somehow wrong. But it won't be. It is Ok to create delegates for other method kinds as well, perhaps language doesn't provide means for that today, but that is not something that this component should worry about in my opinion. It is given a delegate creation to cache, That is all what it is supposed to do, validating the delegate creation is not its job. And knowing what method kind we are dealing with here today doesn't really help because it doesn't affect correctness of what we are doing.

Reply:
> perhaps language doesn't provide means for that today

To me that's an argument *for* an assertion, not against, as we'd likely want to pay attention to this code if this happens.
That said, I don't think this is worth holding the PR over (I'd marked the comment as nit) so I'm okay either way.

Reply:
I am fine with leting @pawchen to decide whether to keep the assert.

Reply:
> @AlekseyTs:  It is Ok to create delegates for other method kinds as well, perhaps language doesn't provide means for that today, but that is not something that this component should worry about.

> @jcouv: we'd likely want to pay attention to this code if this happens.

I agree with both of your opinions. Originally I assumed `Ordinary` here, and all tests we have right now at this point will be `Ordinary` here. In my opinion, BoundDelegateCreation can be from anywhere not in the source, like WinRT Event, and the MethodKind is not limited to ordinary methods.

I think fundamentally the condition supplied for the `if` above is kind of "flawed", what we really care here is, "Are there any type parameters from the target method that we cannot discover simply from it's signature?", as of today, we only observe local functions could potentially answer yes, and we used that to drive down the complexity.

That said, I'm keeping the assert here and I'm also adding comments to capture this discussion.

Reply:
Comments added, please check 84551f1.